cookbook-recipes.md

```diff
@@ -0,0 +1,659 @@
+  const now = new Date();
+  if (jwtCreatedAt > now) {
+    throw new Error("iat value set in future");
+  }
+  if (jwtExpiresAt < now) {
+    throw new Error("JWT is expired");
+  }
+
+  if (audience !== YOUR_EORI) {
+    throw new Error('Wrong audience');
+  }
+
+  // validate the signature (we check with the first certificate in the x5c chain)
+  jwt.verify(clientAssertionJWT, x5cToPem(x5c[0]));
+
+  // validate the certificate chain (is it a chain? is the CA in our list of accepted associations?)
+  let bearerToken = await accessToken(ASSOC_EORI, tokenUrlAssoc);
+  const authenticatedHeader = {
+    "accept": "application/json",
+    "Authorization": "Bearer " + bearerToken
+  };
+
+  let response = await axios.get(trustedUrlAssoc, { headers: authenticatedHeader, params: {} });
+  const trustedList = decodeJWT(response.data.trusted_list_token).trusted_list;
+
+  if (!validateCertificateChain(trustedList, x5c)) {
+    throw new Error("Certificate chain invalid");
+  }
+
+  // contact the association register to see if the client is still in good standing
+
+  // first, get a token
+  bearerToken = await accessToken(ASSOC_EORI, tokenUrlAssoc);
+
+  // then, make the parties call
+
+  const headersParties = { "accept": "application/json", "Authorization": "Bearer " + bearerToken };
+
+  let partiesResponse = await axios.get(partiesUrlAssoc + '/' + clientId, { headers: headersParties, params: {} });
+  let partyToken = partiesResponse.data['party_token'];
+  const decodedPayload = decodeJWTWithHeader(partyToken);
+  let party = decodedPayload["payload"]["party_info"];
+  // check adherence of client
+  checkAdherence(party["adherence"]);
+
+  // generate a token and store it with the expiration date and the client id
+  let uuid = uuidv4();
+  let expiresAt = new Date(new Date().getTime() + 30000);
+  tokenList[uuid] = { clientId: clientId, expiresAt: expiresAt };
+
+  // return the token
+  return uuid;
+}
+
+let party = lookupParty(SP_EORI);
+checkAdherence(party);
+const {arEori, tokenArUrl, delegationArUrl} = extractAuthRegisterFromParty(party);
+bearerToken = await accessToken(arEori, tokenArUrl);
+
+// This should be customized based on the actual API call of the Service Provider
+function createDelegationMask(request) {
+  const policy = {
+    "target": {
+      "resource": {
+        "type": "text",
+        "identifiers": [ "text" ],
+        "attributes": [ "text" ]
+      },
+      "actions": [ "text" ]
+    },
+    "rules": [ { "effect": "text" } ]
+  };
+  let delegationRequest = {
+    "policyIssuer": "text",
+    "target": { "accessSubject": "text" },
+    "policySets": [ { "policies": [ policy ] } ]
+  }
+  return delegationRequest;
+}
+
+async function callApi(token, request) {
+  let clientId = checkToken(token, tokenList);
+  let delegationMask = createDelegationMask(request);
+  let party = lookupParty(SP_EORI);
+  checkAdherence(party);
+  const {arEori, tokenArUrl, delegationArUrl} = extractAuthRegisterFromParty(party);
```

Comment:
Waarom doet de SP een lookup van zijn eigen AR?

Reply:
Waarom niet? Het alternatief is dat je die drie values er hardcoded in zet.